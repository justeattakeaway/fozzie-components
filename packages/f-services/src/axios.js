import axios from 'axios';
import {
    forEach,
    isPlainObject,
    isArray,
    camelCase,
    words,
    upperFirst
} from 'lodash-es';

/**
 * Wrapper for navigator.connection and its Firefox/Safari implementations
 */
const getNetworkDetails = () => {
    if (typeof navigator === 'undefined') return null;

    return navigator.connection
        || navigator.mozConnection
        || navigator.webkitConnection
        || null;
};

/**
 * Returns a timeout in milliseconds based on the current connection speed. Slower connections will have longer timeouts.
 */
const getTimeout = () => {
    const connection = getNetworkDetails();
    const downlink = connection?.downlink;

    const speedSlow2g = 0.25;
    const speed2g = 0.45;
    const speed3g = 0.75;
    const speed4g = 4;

    switch (true) {
        case (downlink < speedSlow2g):
            return 25000;
        case (downlink < speed2g):
            return 20000;
        case (downlink < speed3g):
            return 15000;
        case (downlink < speed4g):
            return 10000;
        case (downlink >= speed4g):
            return 5000;
        default:
            return 10000;
    }
};

/**
 * Converts an objects' property keys into alternate casing using given transformation function.
 *
 * @param {Object} data
 * @param {Function} alternateCaser
 * @returns {Object}
 */
const objectToAlternateCasing = (data, alternateCaser) => {
    const alternateCasedObject = {};
    const alternateCasedArray = [];

    forEach(data, (dataValue, key) => {
        let val = dataValue;

        // checks that a value is a plain object or an array - for recursive key conversion
        if (isPlainObject(val) || isArray(val)) {
            // recursively update keys of any values that are also objects
            val = objectToAlternateCasing(val, alternateCaser);
        }

        if (isArray(data)) {
            alternateCasedArray.push(val);
        } else {
            alternateCasedObject[alternateCaser(key)] = val;
        }
    });

    if (isArray(data)) {
        return alternateCasedArray;
    }

    return alternateCasedObject;
};

/**
 * Converts an object's property names to camelCase
 *
 * @param {Object} data - Object to have its property names converted
 */
const objectToCamelCase = data => objectToAlternateCasing(data, camelCase);

/**
 * Converts a string to Upper-Kebab-Case. Can be used for request headers.
 *
 * @param {string} toConvert
 * @returns {string}
 */
const upperKebabCase = toConvert => words(toConvert)
    .map(upperFirst)
    .join('-');

/**
 * Attach interceptors to the axios client to record response time from an API
 *
 * @param {AxiosInstance} axiosInstance
 * @param {Function} callback - Function to be invoked, takes the [axios response object](https://github.com/axios/axios#response-schema) as a parameter. You should publish your stats (i.e., response times) using this callback.
 */
const setupResponseTimeRecording = ({ interceptors }, callback) => {
    if (!callback || typeof callback !== 'function') return;

    interceptors.request.use(config => {
        config.meta = config.meta || {};
        config.meta.requestStartedAt = new Date().getTime();
        return config;
    });

    interceptors.response.use(response => {
        const timeTakenMs = new Date().getTime() - response.config.meta.requestStartedAt;
        response.responseTimeMs = timeTakenMs; // Extend the default response schema

        callback(response);
        return response;
    });
};

/**
 * Create an axios client.
 *
 * @param {Object} options
 */
const createClient = ({
    baseURL = '',
    headers,
    headerTransform,
    config,
    responseCallback = false
} = {}) => {
    const instance = axios.create({
        baseURL,
        headers: headerTransform
            ? headerTransform(headers)
            : objectToAlternateCasing(headers, upperKebabCase),
        timeout: getTimeout(),
        ...config
    });

    setupResponseTimeRecording(instance, responseCallback);

    return instance;
};

/**
 * Wrapper for createClient that transforms response objects' property names to camelCase.
 *
 * @param {string} options
 */
const createCamelCaseClient = ({
    baseURL,
    headers,
    headerTransform
} = {
    baseURL: ''
}) => createClient({
    baseURL,
    headers,
    headerTransform,
    config: {
        transformResponse: [
            ...axios.defaults.transformResponse,
            data => objectToCamelCase(data)
        ]
    }
});

export default {
    createClient,
    createCamelCaseClient,
    objectToCamelCase,
    getNetworkDetails,
    setupResponseTimeRecording
};
