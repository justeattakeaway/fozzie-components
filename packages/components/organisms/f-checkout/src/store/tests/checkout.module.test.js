import axios from 'axios';
import CheckoutModule from '../checkout.module';
import checkoutDelivery from '../../demo/checkout-delivery.json';
import basketDelivery from '../../demo/get-basket-delivery.json';
import checkoutAvailableFulfilment from '../../demo/checkout-available-fulfilment.json';
import customerAddresses from '../../demo/get-address.json';
import geoLocationDetails from '../../demo/get-geo-location.json';

import {
    UPDATE_AUTH,
    UPDATE_AVAILABLE_FULFILMENT_TIMES,
    UPDATE_BASKET_DETAILS,
    UPDATE_CUSTOMER_DETAILS,
    UPDATE_FULFILMENT_ADDRESS,
    UPDATE_FULFILMENT_TIME,
    UPDATE_IS_FULFILLABLE,
    UPDATE_ISSUES,
    UPDATE_STATE,
    UPDATE_USER_NOTE,
    UPDATE_GEO_LOCATION
} from '../mutation-types';

const { actions, mutations } = CheckoutModule;

const {
    createGuestUser,
    getAddress,
    getAvailableFulfilment,
    getBasket,
    getCheckout,
    getGeoLocation,
    updateCheckout,
    setAuthToken,
    updateAddressDetails,
    updateCustomerDetails,
    updateFulfilmentTime,
    updateUserNote
} = actions;

const mobileNumber = '+447111111111';

const customerDetails = {
    mobileNumber
};

const authToken = 'sampleToken';

const address = {
    line1: 'line 1',
    line2: 'line 2',
    city: 'city',
    postcode: 'postcode'
};

const locationData = {
    addressLines: [
        '1 Jazz Avenue',
        'Strange Town',
        'JZ1 1AA'
    ]
};

const time = {
    from: 'fromTime',
    to: 'toTime'
};

const isFulfillable = false;

const issues = [
    {
        code: 'RESTAURANT_UNAVAILABLE'
    }
];

const userNote = 'Beware of the dachshund';

const defaultState = {
    id: '',
    serviceType: '',
    restaurantId: '',
    basket: {
        id: '',
        total: 0
    },
    customer: {
        firstName: '',
        lastName: '',
        email: '',
        mobileNumber: ''
    },
    orderId: '',
    time: {
        from: '',
        to: ''
    },
    address: {
        line1: '',
        line2: '',
        city: '',
        postcode: ''
    },
    isFulfillable: true,
    issues: [],
    notices: [],
    messages: [],
    availableFulfilment: {
        times: [],
        isAsapAvailable: false
    },
    authToken: '',
    isLoggedIn: false,
    userNote: '',
    geolocation: null
};

let state = CheckoutModule.state();

describe('CheckoutModule', () => {
    it('should create default state when initialised.', () => {
        // Assert
        expect(state).toEqual(defaultState);
    });

    describe('mutations ::', () => {
        beforeEach(() => {
            state = defaultState;
        });

        describe(`${UPDATE_STATE} ::`, () => {
            it('should update state with delivery response.', () => {
                // Act
                mutations[UPDATE_STATE](state, checkoutDelivery);

                // Assert
                expect(state).toMatchSnapshot();
            });

            it('should leave customer state empty if no customer data is returned from the API.', () => {
                // Arrange
                checkoutDelivery.customer = null;

                // Act
                mutations[UPDATE_STATE](state, checkoutDelivery);

                // Assert
                expect(state.customer).toEqual(defaultState.customer);
            });

            it('should leave address state empty if no address data is returned from the API.', () => {
                // Arrange
                checkoutDelivery.address = null;

                // Act
                mutations[UPDATE_STATE](state, checkoutDelivery);

                // Assert
                expect(state.address).toEqual(defaultState.address);
            });
        });

        describe(`${UPDATE_AUTH} ::`, () => {
            it('should update state with authToken and set `isLoggedIn` to true', () => {
                // Act
                mutations[UPDATE_AUTH](state, authToken);

                // Assert
                expect(state.authToken).toEqual(authToken);
                expect(state.isLoggedIn).toBeTruthy();
            });
        });

        describe(`${UPDATE_AVAILABLE_FULFILMENT_TIMES} ::`, () => {
            it('should update state with the availableFulfilment response', () => {
                // Arrange
                const expectedAvailableFulfilmentTimes = checkoutAvailableFulfilment.times;

                // Act
                mutations[UPDATE_AVAILABLE_FULFILMENT_TIMES](state, checkoutAvailableFulfilment);

                // Assert
                expect(state.availableFulfilment.times).toEqual(expectedAvailableFulfilmentTimes);
                expect(state.availableFulfilment.isAsapAvailable).toBe(true);
            });
        });

        describe(`${UPDATE_CUSTOMER_DETAILS} ::`, () => {
            it('should update state with received value', () => {
                // Arrange & Act
                mutations[UPDATE_CUSTOMER_DETAILS](state, { mobileNumber });

                // Assert
                expect(state.customer.mobileNumber).toEqual(mobileNumber);
            });
        });

        describe(`${UPDATE_BASKET_DETAILS} ::`, () => {
            it('should update state with received value', () => {
                // Arrange & Act
                const eventData = {
                    serviceType: 'delivery',
                    basket: {
                        id: '11111',
                        total: 12.50
                    },
                    restaurantId: '22222'
                };
                mutations[UPDATE_BASKET_DETAILS](state, eventData);

                // Assert
                expect(state.serviceType).toEqual(eventData.serviceType);
                expect(state.basket).toEqual(eventData.basket);
                expect(state.restaurant).toEqual(eventData.restaurant);
            });
        });

        describe(`${UPDATE_GEO_LOCATION} ::`, () => {
            it('should update state with received values', () => {
                // Arrange (Long / Lat)
                const geometryData = [-0.10358, 51.51469];

                // Act
                mutations[UPDATE_GEO_LOCATION](state, geometryData);

                // Assert
                expect(state.geolocation).toEqual({
                    latitude: geometryData[1],
                    longitude: geometryData[0]
                });
            });
        });

        it.each([
            [UPDATE_FULFILMENT_ADDRESS, 'address', address],
            [UPDATE_FULFILMENT_TIME, 'time', time],
            [UPDATE_IS_FULFILLABLE, 'isFulfillable', isFulfillable],
            [UPDATE_ISSUES, 'issues', issues],
            [UPDATE_USER_NOTE, 'userNote', userNote]
        ])('%s :: should update state with received value', (mutationName, propertyName, propertyValue) => {
            // Arrange & Act
            mutations[mutationName](state, propertyValue);

            // Assert
            expect(state[propertyName]).toEqual(propertyValue);
        });
    });

    describe('actions ::', () => {
        let commit;
        let dispatch;

        const payload = {
            url: 'http://localhost/account/checkout',
            tenant: 'uk',
            language: 'en-GB',
            timeout: '1000',
            postData: null
        };

        beforeEach(() => {
            commit = jest.fn();
            dispatch = jest.fn();
        });

        describe('getCheckout ::', () => {
            let config;

            beforeEach(() => {
                config = {
                    headers: {
                        'Content-Type': 'application/json',
                        Authorization: `Bearer ${state.authToken}`
                    },
                    timeout: payload.timeout
                };

                axios.get = jest.fn(() => Promise.resolve({ data: checkoutDelivery }));
            });

            it(`should get the checkout details from the backend and call ${UPDATE_STATE} mutation.`, async () => {
                // Act
                await getCheckout({ commit, state, dispatch }, payload);

                // Assert
                expect(axios.get).toHaveBeenCalledWith(payload.url, config);
                expect(commit).toHaveBeenCalledWith(UPDATE_STATE, checkoutDelivery);
            });

            it('should call `analytics/updateAutofill` mutation with an array of updated field names.', async () => {
                // Act
                await getCheckout({ commit, state, dispatch }, payload);

                // Assert
                expect(dispatch).toHaveBeenCalledWith('analytics/updateAutofill', state, { root: true });
            });
        });

        describe('getBasket ::', () => {
            it(`should get the basket details from the backend and call ${UPDATE_BASKET_DETAILS} mutation.`, async () => {
                // Arrange
                const config = {
                    headers: {
                        'Content-Type': 'application/json',
                        'Accept-Tenant': payload.tenant,
                        'Accept-Language': payload.language
                    },
                    timeout: payload.timeout
                };

                axios.get = jest.fn(() => Promise.resolve({ data: basketDelivery }));

                // Act
                await getBasket({ commit, state, dispatch }, payload);

                // Assert
                expect(axios.get).toHaveBeenCalledWith(payload.url, config);
                expect(commit).toHaveBeenCalledWith(UPDATE_BASKET_DETAILS, {
                    serviceType: basketDelivery.ServiceType.toLowerCase(),
                    restaurantId: basketDelivery.RestaurantId,
                    basket: {
                        id: basketDelivery.BasketId,
                        total: basketDelivery.BasketSummary.BasketTotals.Total
                    }
                });
            });
        });

        describe('getAddress ::', () => {
            it(`should get the address details from the backend and call ${UPDATE_FULFILMENT_ADDRESS} mutation.`, async () => {
                // Arrange
                const config = {
                    headers: {
                        'Content-Type': 'application/json',
                        'Accept-Language': payload.language,
                        Authorization: `Bearer ${state.authToken}`
                    },
                    timeout: payload.timeout
                };

                axios.get = jest.fn(() => Promise.resolve({ data: customerAddresses }));
                const [expectedAddress] = customerAddresses.Addresses;

                // Act
                await getAddress({ commit, state, dispatch }, payload);

                // Assert
                expect(axios.get).toHaveBeenCalledWith(payload.url, config);
                expect(commit).toHaveBeenCalledWith(UPDATE_FULFILMENT_ADDRESS, {
                    line1: expectedAddress.Line1,
                    line2: expectedAddress.Line2,
                    city: expectedAddress.City,
                    postcode: expectedAddress.ZipCode
                });
            });

            it('should call `analytics/updateAutofill` mutation with an array of updated field names.', async () => {
                // Act
                await getAddress({ commit, state, dispatch }, payload);

                // Assert
                expect(dispatch).toHaveBeenCalledWith('analytics/updateAutofill', state, { root: true });
            });
        });

        describe('updateCheckout ::', () => {
            payload.data = {
                mobileNumber
            };

            let config;

            beforeEach(() => {
                config = {
                    headers: {
                        'Content-Type': 'application/json',
                        Authorization: `Bearer ${authToken}`
                    },
                    timeout: payload.timeout
                };
                axios.patch = jest.fn(() => Promise.resolve({
                    status: 200,
                    data: {}
                }));
            });

            it('should post the checkout details to the backend.', async () => {
                // Act
                await updateCheckout({ commit, state }, payload);

                // Assert
                expect(axios.patch).toHaveBeenCalledWith(payload.url, payload.data, config);
            });
        });

        describe('createGuestUser ::', () => {
            let config;
            payload.url = 'http://localhost/account/createguest';
            payload.data = {
                firstName: 'Joe',
                lastName: 'Bloggs',
                email: 'joe@test.com'
            };

            beforeEach(() => {
                config = {
                    headers: {
                        'Content-Type': 'application/json',
                        'Accept-Tenant': payload.tenant
                    },
                    timeout: payload.timeout
                };

                axios.post = jest.fn(() => Promise.resolve({
                    status: 200,
                    data: {
                        token: 'otacToken',
                        type: 'otac'
                    }
                }));
            });

            it('should post the create guest user request to the backend.', async () => {
                // Act
                await createGuestUser({ commit, state }, payload);

                // Assert
                expect(axios.post).toHaveBeenCalledWith(payload.url, payload.data, config);
            });
        });

        describe('getAvailableFulfilment ::', () => {
            let config;

            beforeEach(() => {
                config = {
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    timeout: payload.timeout
                };

                axios.get = jest.fn(() => Promise.resolve({ data: checkoutAvailableFulfilment }));
            });

            it(`should get the checkout available fulfilment details from the backend and call ${UPDATE_AVAILABLE_FULFILMENT_TIMES} mutation.`, async () => {
                // Act
                await getAvailableFulfilment({ commit }, payload);

                // Assert
                expect(axios.get).toHaveBeenCalledWith(payload.url, config);
                expect(commit).toHaveBeenCalledWith(UPDATE_AVAILABLE_FULFILMENT_TIMES, checkoutAvailableFulfilment);
            });
        });

        describe('updateFulfilmentTime ::', () => {
            it(`should call ${UPDATE_FULFILMENT_TIME} mutation.`, () => {
                // Act
                updateFulfilmentTime({ commit, state }, time);

                // Assert
                expect(commit).toHaveBeenCalledWith(UPDATE_FULFILMENT_TIME, time);
            });
        });

        describe('getGeoLocation ::', () => {
            // Arrange
            let config;
            payload.postData = locationData;

            beforeEach(() => {
                config = {
                    headers: {
                        'Content-Type': 'application/json',
                        Authorization: `Bearer ${authToken}`
                    },
                    timeout: payload.timeout
                };

                axios.post = jest.fn(() => Promise.resolve({
                    status: 200,
                    data: geoLocationDetails
                }));
            });

            describe('if the user is logged in', () => {
                beforeEach(() => {
                    state.isLoggedIn = true;
                });

                it(`should get the geo location details from the backend and call ${UPDATE_GEO_LOCATION} mutation.`, async () => {
                    // Act
                    await getGeoLocation({ commit, state }, payload);

                    // Assert
                    expect(axios.post).toHaveBeenCalledWith(payload.url, locationData, config);
                    expect(commit).toHaveBeenCalledWith(UPDATE_GEO_LOCATION, geoLocationDetails.geometry.coordinates);
                });
            });

            describe('if the user is not logged in', () => {
                beforeEach(() => {
                    state.isLoggedIn = false;
                });

                it('should not make api call and should not call mutation.', async () => {
                    // Act
                    await getGeoLocation({ commit, state }, payload);

                    // Assert
                    expect(axios.post).not.toHaveBeenCalled();
                    expect(commit).not.toHaveBeenCalled();
                });
            });
        });

        it.each([
            [setAuthToken, UPDATE_AUTH, authToken],
            [updateAddressDetails, UPDATE_FULFILMENT_ADDRESS, address],
            [updateCustomerDetails, UPDATE_CUSTOMER_DETAILS, customerDetails],
            [updateUserNote, UPDATE_USER_NOTE, userNote]
        ])('%s should call %s mutation with passed value', (action, mutation, value) => {
            // Act
            action({ commit, dispatch }, value);

            // Assert
            expect(commit).toHaveBeenCalledWith(mutation, value);
        });

        it.each([
            [updateAddressDetails, address],
            [updateCustomerDetails, customerDetails]
        ])('%s should dispatch `analytics/updateChangedFields` action with first key of passed value', (action, value) => {
            // Act
            action({ commit, dispatch }, value);

            const [field] = Object.keys(value);

            // Assert
            expect(dispatch).toHaveBeenCalledWith('analytics/updateChangedField', field, { root: true });
        });
    });
});
