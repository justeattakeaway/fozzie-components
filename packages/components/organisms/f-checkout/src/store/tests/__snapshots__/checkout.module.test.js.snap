// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutModule mutations :: UPDATE_STATE :: should update state with delivery response. 1`] = `
Object {
  "address": Object {
    "city": "Bristol",
    "line1": "1 Bristol Road",
    "line2": "Flat 1",
    "postcode": "BS1 1AA",
  },
  "authToken": "",
  "availableFulfilment": Object {
    "isAsapAvailable": false,
    "times": Array [],
  },
  "basket": Object {
    "id": "",
    "total": 0,
  },
  "customer": Object {
    "email": "",
    "firstName": "Joe",
    "lastName": "",
    "mobileNumber": "+447111111111",
  },
  "geolocation": null,
  "id": "12345",
  "isFulfillable": true,
  "isLoggedIn": false,
  "issues": Array [],
  "messages": Array [
    Object {
      "message": Object {
        "text": "Please hurry, the restaurant is closing soon",
      },
      "type": "warning",
    },
    Object {
      "message": Object {
        "text": "We're sorry, some items in your basket are no longer available",
      },
      "type": "information",
    },
  ],
  "notices": Array [
    Object {
      "notice": Object {
        "text": "If you have a food allergy or intolerance (or someone you're ordering for has), <a href=\\"https://greggs.co.uk/nutrition\\" data-test-id=\\"allergen-url-link\\" target=\\"_blank\\" rel=\\"noopener\\">read what this restaurant has to say about allergies</a> before placing your order. Do not order if you cannot get the allergy information you need.",
      },
      "type": "allergy",
    },
  ],
  "orderId": "",
  "restaurantId": "",
  "serviceType": "delivery",
  "time": Object {
    "from": "2020-01-01T00:00+00:00",
    "to": "2020-01-01T00:00+00:00",
  },
  "userNote": "",
}
`;
